### PR TITLE
Add NonderterministicFastCommit to speed up migrations when ordering isn't required

### PR DESCRIPTION
**Pull request by fxamacker**
Updates #394

This PR adds `NonderterministicFastCommit` to improve speed, bytes/op, and allocs/op compared to `FastCommit`.

`NonderterministicFastCommit` commits changes in nondeterministic order. It can be used by migration program when ordering isn't required.

#### FastCommit vs NonderterministicFastCommit

Casual micro benchstat output (from my busy desktop):

```
                             │  before.txt  │              after.txt               │
                             │    sec/op    │    sec/op     vs base                │
StorageFastCommit/10-12         89.72µ ± 4%   57.50µ ±  3%  -35.92% (p=0.000 n=10)
StorageFastCommit/100-12        118.9µ ± 1%   116.0µ ±  4%        ~ (p=0.436 n=10)
StorageFastCommit/1000-12       4.086m ± 5%   2.397m ± 25%  -41.35% (p=0.000 n=10)
StorageFastCommit/10000-12     12.629m ± 4%   9.857m ±  3%  -21.95% (p=0.000 n=10)
StorageFastCommit/100000-12    102.73m ± 0%   72.26m ±  1%  -29.66% (p=0.000 n=10)
StorageFastCommit/1000000-12     1.544 ± 2%    1.141 ±  2%  -26.09% (p=0.000 n=10)
geomean                         6.661m        4.848m        -27.21%

                             │  before.txt  │              after.txt              │
                             │     B/op     │     B/op      vs base               │
StorageFastCommit/10-12        28.92Ki ± 0%   28.05Ki ± 0%  -3.00% (p=0.000 n=10)
StorageFastCommit/100-12       286.4Ki ± 0%   278.6Ki ± 0%  -2.71% (p=0.000 n=10)
StorageFastCommit/1000-12      3.009Mi ± 0%   2.901Mi ± 0%  -3.58% (p=0.000 n=10)
StorageFastCommit/10000-12     28.65Mi ± 0%   27.79Mi ± 0%  -2.98% (p=0.000 n=10)
StorageFastCommit/100000-12    278.8Mi ± 0%   271.1Mi ± 0%  -2.75% (p=0.000 n=10)
StorageFastCommit/1000000-12   2.923Gi ± 0%   2.821Gi ± 0%  -3.49% (p=0.000 n=10)
geomean                        9.101Mi        8.820Mi       -3.09%

                             │ before.txt  │             after.txt              │
                             │  allocs/op  │  allocs/op   vs base               │
StorageFastCommit/10-12         219.0 ± 0%    205.0 ± 0%  -6.39% (p=0.000 n=10)
StorageFastCommit/100-12       1.980k ± 0%   1.875k ± 0%  -5.30% (p=0.000 n=10)
StorageFastCommit/1000-12      19.23k ± 0%   18.23k ± 0%  -5.22% (p=0.000 n=10)
StorageFastCommit/10000-12     191.1k ± 0%   181.1k ± 0%  -5.24% (p=0.000 n=10)
StorageFastCommit/100000-12    1.918M ± 0%   1.816M ± 0%  -5.30% (p=0.000 n=10)
StorageFastCommit/1000000-12   19.15M ± 0%   18.15M ± 0%  -5.22% (p=0.000 n=10)
geomean                        62.31k        58.91k       -5.45%

```
______

- [ ] Targeted PR against `main` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/atree/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

**Comment by fxamacker** [comment]
Looks like validation and storage health checks passed tonight with this PR added to atree migration program at:
- (locally modified) https://github.com/onflow/flow-go/tree/feature/atree-inlining-cadence-v0.42

Need to take another look at logs to confirm.

File: storage.go [hunk 1]
```diff
@@ -964,6 +969,204 @@ func (s *PersistentSlabStorage) FastCommit(numWorkers int) error {
 	return nil
 }
 
+// NonderterministicFastCommit commits changes in nondeterministic order.
+// This is used by migration program when ordering isn't required.
+func (s *PersistentSlabStorage) NonderterministicFastCommit(numWorkers int) error {
```
Comment (turbolent) [inline storage.go:974]: Would it be possible to get a diff of this new function and the existing deterministic version of it? It would make it easier to review, by seeing how they differ.

If the functions only differ in a small way, maybe it's possible to extract the shared code and only make the parts that differ configurable?

**Review by turbolent — COMMENTED** [review]
Great idea and great work!

**Comment by fxamacker** [comment]
> Would it be possible to get a diff of this new function and the existing deterministic version of it? It would make it easier to review, by seeing how they differ.
>
> If the functions only differ in a small way, maybe it's possible to extract the shared code and only make the parts that differ configurable?

@turbolent  Good suggestions!  However, in this case there are more changes than unchanged code so a diff, etc. isn't as helpful as it normally would be.

The changes are mostly about eliminating overhead and parallelizing as much as possible.  

I outlined steps in both functions below so we can see the differences at higher level.

`FastCommit()` commits deltas in deterministic order:
- sort non-temp slab IDs by address and index
- send non-temp slab IDs to job queue
- (in goroutines) read and encode slabs from deltas
  - job queue (input): slab ID
  - result queue (output): encoded slab and possible error
- collect all encoded results from result queue in temporary map
- commit deleted and encoded results in order to base storage

In order to commit in deterministic order, the commit step has to wait until all slabs are encoded, and there are also overheads from sorting slab IDs and collecting encoded results in temporary map.

`NonderterministicFastCommit()` commits deltas in nondeterministic order:
- send non-temp and non-empty slab to job queue
- (in goroutines) read and encode slabs 
  - job queue (input): slab ID and slab
  - result queue (output): encoded slab and possible error
- commit deleted slabs to base storage  
- commit encoded slabs from result queue to base storage

In nondeterministic approach, the commit steps are in parallel with encoding and there isn't any temp map to collect data.

Even though goroutine encoding step is similar, `FastCommit()` reads slab from `deltas` map while `NonderterministicFastCommit()` reads slab from job queue.  Although I can modify `FastCommit()` to be the same in this aspect, it doesn't seem worthwhile since that would use more memory without making `FastCommit()` faster.

**Comment by turbolent** [comment]
@fxamacker Thank you for the description! 👌  I'll take another look today

File: storage.go [hunk 2]
```diff
@@ -964,6 +969,204 @@ func (s *PersistentSlabStorage) FastCommit(numWorkers int) error {
...
+		// Ignore slabs not owned by accounts
+		if k.address == AddressUndefined {
+			continue
+		}
+		if v == nil {
+			deletedSlabCount++
+		} else {
+			modifiedSlabCount++
+		}
+	}
```
Comment (ramtinms) [inline storage.go:1044]: Just a raw idea, anyway we can capture this data as part of updates, so we can skip iteration here.

**Review by ramtinms — APPROVED** [review]
Nice work, looks good to me, thanks for adding the tests to compare encoded final results.

File: storage.go (hunk 2 above)
Comment (fxamacker) [reply storage.go]: > Just a raw idea, anyway we can capture this data as part of updates, so we can skip iteration here.

@ramtinms I like the idea of skipping iterations if possible!

By "updates", do you mean in `Store()` when modified or deleted slabs are stored in `deltas`?  If so, it may be less memory efficient for non-migration use cases.

To reduce iterations here in `NonderterministicFastCommit()`, I pushed commit a2a4f5c08636bd27e3c436814dff3c7db1da2db5 to iterate only once to get modified and deleted slabs to process later.

File: .github/workflows/safer-golangci-lint.yml [hunk 3]
```diff
@@ -64,12 +32,12 @@ jobs:
       contents: read
     steps:
       - name: Checkout source
-        uses: actions/checkout@v3
+        uses: actions/checkout@0ad4b8fadaa221de15dcec353f45205ec38ea70b # v4.1.4
```
Comment (turbolent) [inline .github/workflows/safer-golangci-lint.yml:35]: Are these dependency updates necessary?

File: storage.go [hunk 4]
```diff
@@ -964,6 +969,202 @@ func (s *PersistentSlabStorage) FastCommit(numWorkers int) error {
...
+	// Modified slabs need to be encoded (in parallel) and stored in underlying storage.
+	modifiedSlabCount := 0
+	// Deleted slabs need to be removed from underlying storage.
+	deletedSlabCount := 0
+	for k, v := range s.deltas {
+		// Ignore slabs not owned by accounts
+		if k.address == AddressUndefined {
+			continue
+		}
+		if v == nil {
```
Comment (turbolent) [inline storage.go:1045]: Maybe use `len(v) == 0` here?

File: storage.go [hunk 5]
```diff
@@ -964,6 +969,202 @@ func (s *PersistentSlabStorage) FastCommit(numWorkers int) error {
...
+			index := len(slabIDsWithOwner) - 1 - deletedSlabCount
```
Comment (turbolent) [inline storage.go:1046]: It took me a moment to realize that this loop fills modified slabs from the front, and here, deleted slabs from the end. Maybe add a comment

**Review by turbolent — APPROVED** [review]
Great work!

I think I understand what the new function does, but it's quite hard to follow given the many channels that are involved. Maybe https://pkg.go.dev/golang.org/x/sync/errgroup can help here?

File: .github/workflows/safer-golangci-lint.yml (hunk 3 above)
Comment (fxamacker) [reply .github/workflows/safer-golangci-lint.yml]: These updates to pin dependencies improves OpenSSF score.  However, I reverted the commit that updated golangci-lint to 1.53.3 (out of scope of this PR).

Example when all dependencies are pinned (score 0-10 with 10 being perfect):

image

For more, replace "example/repo" in this URL to see complete report (not all repos are scored yet).
- https://scorecard.dev/viewer/?uri=github.com/example/repo

**Comment by fxamacker** [comment]
> @fxamacker Does this also have to get ported to the 1.0 branch?

@turbolent I will port it for completeness, but it is only needed if we use the optimization for Cadence 1.0 migration without atree inlining.
